amethyst_network/src/error.rs

```diff
@@ -0,0 +1,85 @@
+//! Module containing error handling logic.
+
+use std::{
+    error::Error as StdError,
+    fmt::{Display, Formatter, Result as FmtResult},
+    io,
+    sync::mpsc,
+};
+
+use crate::server::ServerSocketEvent;
+
+/// The `amethyst_network` result type.
+pub type NetworkError<T> = Result<T, NetworkErrorKind>;
+
+/// Wrapper for all errors who could occur in `amethyst_network`.
+#[derive(Debug)]
+pub enum NetworkErrorKind {
+    /// Error that could occur on the UDP-socket
+    UdpError(laminar::error::NetworkError),
+    /// Error that could occur whit IO.
+    IoError(io::Error),
+    /// Error that could occur when serializing whit `bincode`
+    SerializeError(bincode::ErrorKind),
+    /// Error that could occur when sending an `ServerSocketEvent` to some channel.
+    ChannelSendError(mpsc::SendError<ServerSocketEvent>),
+}
+
+impl StdError for NetworkErrorKind {
+    fn description(&self) -> &str {
+        match *self {
+            NetworkErrorKind::UdpError(_) => "An UDP error occurred",
+            NetworkErrorKind::IoError(ref error) => error.description(),
+            NetworkErrorKind::ChannelSendError(ref error) => error.description(),
+            NetworkErrorKind::SerializeError(ref error) => error.description(),
+        }
+    }
+
+    fn cause(&self) -> Option<&dyn StdError> {
+        match *self {
+            NetworkErrorKind::IoError(ref error) => Some(error),
+            NetworkErrorKind::SerializeError(ref error) => Some(error),
+            NetworkErrorKind::ChannelSendError(ref error) => Some(error),
+            _ => None,
+        }
+    }
+}
+
+impl Display for NetworkErrorKind {
+    fn fmt(&self, fmt: &mut Formatter<'_>) -> FmtResult {
+        match *self {
+            NetworkErrorKind::UdpError(ref error) => write!(fmt, "UDP-error occurred {}", error),
+            NetworkErrorKind::IoError(ref error) => write!(fmt, "IO-error occurred {}", error),
+            NetworkErrorKind::ChannelSendError(ref error) => {
+                write!(fmt, "Channel send error occurred  {}", error)
+            }
+            NetworkErrorKind::SerializeError(ref error) => {
+                write!(fmt, "Channel send error occurred  {}", error)
+            }
+        }
+    }
+}
+
+impl From<io::Error> for NetworkErrorKind {
+    fn from(e: io::Error) -> NetworkErrorKind {
+        NetworkErrorKind::IoError(e)
+    }
+}
+
+impl From<mpsc::SendError<ServerSocketEvent>> for NetworkErrorKind {
+    fn from(e: mpsc::SendError<ServerSocketEvent>) -> NetworkErrorKind {
+        NetworkErrorKind::ChannelSendError(e)
+    }
+}
+
+impl From<laminar::error::NetworkError> for NetworkErrorKind {
+    fn from(e: laminar::error::NetworkError) -> NetworkErrorKind {
+        NetworkErrorKind::UdpError(e)
+    }
+}
+
+impl From<Box<bincode::ErrorKind>> for NetworkErrorKind {
```

Comment:
No this could not be done. `bincode::deserialize` is returning an error of type: `Result<T, Box<ErrorKind>`. I need to have a conversion from that to my `NetworkErrorKind`.